src/dynarec/dynarec_arm_d9.c

```diff
@@ -281,39 +281,100 @@ uintptr_t dynarecD9(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int ninst,
             CALL(arm_f2xm1, -1, 0);
             #else
             v1 = x87_get_st(dyn, ninst, x1, x2, 0, NEON_CACHE_ST_D);
-            //if(ST0.d!=0.0)
-            //    ST0.d = exp2(ST0.d)-1.0;
-            VMOV_64(0, v1);
-            CALL_1D(exp2, 0);   // return is d0
+            if(!box86_dynarec_fastround) {
+                LDRH_IMM8(x1, xEmu, offsetof(x86emu_t, cw));    // hopefully cw is not too far for an imm8
+                UBFX(x1, x1, 10, 2);    // extract round...
+                UBFX(x2, x1, 1, 1);     // swap bits 0 and 1
+                BFI(x2, x1, 1, 1);
+                VMRS(x14);               // get fpscr
+                MOV_REG(x3, x14);
+            }
             if((PK(0)==0xD9 && PK(1)==0xE8) && // next inst is FLD1
             (PK(2)==0xDE && PK(3)==0xC1)) {
                 MESSAGE(LOG_DUMP, "Hack for fld1 / faddp st1, st0\n");
```

Comment:
I don't understand what you are trying to do here.

Reply:
Line 284 \~ line 291 copy fpscr to x14 and x3. X14 is used to save the original value and be pushed when the function is called, and x3 is to be set to the new value.
```c
            if(!box86_dynarec_fastround) {
                LDRH_IMM8(x1, xEmu, offsetof(x86emu_t, cw));    // hopefully cw is not too far for an imm8
                UBFX(x1, x1, 10, 2);    // extract round...
                UBFX(x2, x1, 1, 1);     // swap bits 0 and 1
                BFI(x2, x1, 1, 1);
                VMRS(x14);               // get fpscr
                MOV_REG(x3, x14);
            }
```

Line 295 ~ line 303 apply the rounding mode of x87emu, but changing in some situations, like https://github.com/ptitSeb/box86/pull/951/files/9814aefa576ddd5e97051d1cf2081460b4d431d1#r1564733419.
```c

                if(!box86_dynarec_fastround) {
                    VCMP_F64_0(v1);
                    B_MARK(cGE);           // if ST0 < 0 and if the rounding mode is toward 0, then use upward
                    TSTS_IMM8(x2, 0b01);
                    AND_IMM8_COND(cNE, x2, x2, 0b01); // 11 (TOWARDZERO) -> 01 (UPWARD), 01 -> 01
                    MARK;
                    BFI(x3, x2, 22, 2);     // inject new round
                    VMSR(x3);
                }
```

Reply:
When it is not optimizing for `FLD1; FADDP ST1, ST0;`, `expm1(ln2*ST0)` instead of `exp2(ST0) - 1` is used for better precision.